### PR TITLE
Use `fgets()` for reading from `stdin` and a temporay buffer to read from file

### LINKED ISSUE #4015: Bug: `read()` gives invalid output
This bug was caught when working on #4011.

> b.f90
```fortran
program main
    character :: c
    open(11, file="intrinsics_60_data.txt", form="formatted", access="stream", status="old")
    read(11, *) c
    print *, c
    print*, len(c)
end
```
```console
$ gfortran b.f90 && ./a.out
 H
           1
$ lfortran b.f90 --no-warnings
Hello
5
```

> intrinsics_60_data.txt
```
Hello
```

When ready with a fix, please also verify the integration test `intrinsics_60.f90`.

### PR DESCRIPTION
**Pull request by kmr-srbh**
fixes #4015 

Use `fgets()` to read `n` characters from `stdin` where `n` is the length of the variable which will store the data. When reading from a file, we use a temporary buffer to store the whole data. Post this, we add the required chunk of data to the main buffer.
 
### Console read
```fortran
program main
    character(len=2) :: c
    read *, c
    print *, c
    print*, len(c)
end
```
#### Output on main
```console
apple
5
```
#### Output on branch
```console
ap
2
```

### File read
```fortran
program hello
    character(3) :: c
    open(11, file="hello.txt", form="formatted", access="stream", status="old")
    read(11, *) c
    print *, c
    print *, len(c)
end program
```
Content of `hello.txt` is
```
Hello

```
#### Output on main
```console
Hello
5
```
#### Output on branch
```console
Hel
3
```

### CODE REVIEW

**Comment by kmr-srbh** [comment]
```console
The following tests FAILED:
        282 - intrinsics_60 (Failed)
        669 - file_03 (Failed)
```

`intrinsics_60.f90` fails because we are now reading whitespace characters also. I see that GFortran does not read whitespace characters. We can fix this by doing a char-by-char read. I read online that it can be slow. @certik could you please guide me on this?

`file_03.f90` fails unexpectedly! All the other file tests pass, but this one fails.

**Comment by certik** [comment]
Tests have to pass.

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -2383,10 +2383,11 @@ LFORTRAN_API void _lfortran_read_array_int32(int32_t *p, int array_size, int32_t
 
 LFORTRAN_API void _lfortran_read_char(char **p, int32_t unit_num)
```
Comment (gxyd) [inline src/libasr/runtime/lfortran_intrinsics.c:2384]: I think we should be obtaining the "char_length" at compile time by passing another argument here to `_lfortran_read_char`.
```f90
program main
    character :: c
    open(11, file="intrinsics_60_data.txt", form="formatted", access="stream", status="old")
    read(11, *) c
    print *, c
    print*, len(c)
end
```

from this we already know that `c` has a length of `1`, which is know at compile-time, if we can pass that as an argument to `_llfortran_read_char` then we know better about how to read data.

Comment (kmr-srbh) [reply src/libasr/runtime/lfortran_intrinsics.c:2384]: We get the character length for read through the length of the passed variable.
```c
int n = strlen(*p);
```

Comment (gxyd) [reply src/libasr/runtime/lfortran_intrinsics.c:2384]: Ah yes, I think you are right. Sorry, I haven't been able to give this a look.

**Comment by Shaikh-Ubaid** [comment]
The approach used in this PR seems fine to me. @kmr-srbh you just need to get the tests to pass. Just strip down the failing tests (for example `intrinsics_60` or `file_03`) to the smallest code that fails. Now you have a minimum reproducible example. Use it to find a fix. Ensure your fix does not break any other tests. And that's it!

**Comment by kmr-srbh** [comment]
We now read files character by character skipping whitespaces. However, the integration tests `file_01.f90` and `file_03.f90` fail with an `EOF error`. I am handling EOF in the implementation, but still the error occurs. Also, only `file_03.f90` had failed before, but we have `file_01.f90` failing now.

**Comment by kmr-srbh** [comment]
This implementation is now complete.

**Comment by kmr-srbh** [comment]
The Windows test fails unexpectedly.

**Comment by certik** [comment]
The Windows build error must be investigated.

**Comment by kmr-srbh** [comment]
@certik this PR is now ready.

**Review by certik — APPROVED** [review]
I think that this is fine, thanks!

**Pull request merged**